### ISSUE TITLE
There was no increase in speed after installing flash-attn and xformer.

### ISSUE THREAD

**Issue by kongwanbianjinyu** [FoundationVision/VAR#59]
Hello,

I installed the flash-attn and xformer packages to accelerate the training process, but noticed no speed improvement compared to slow_attn. I'm using 4 RTX 6000 GPUs to train VAR-d16 for 200 epochs, with a global_batch_size of 96 due to memory constraints. The whole training process would take roughly 1000 hours. Could you kindly share your hardware settings and the time required to train VAR-d16?

Best,
Jiachen

**Comment by Kumbong** [comment]
Flash attention does not support using attention masks so the code falls back to xformers or slow_attn (which is basically torch.sdpa). xformers uses mem efficient attention which I think is the [same algorithm that is called by torch.sdpa under the hood when you provide a mask](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) so you should see similar speed when using former and torch.sdpa.   mem_efficient attention reduces the memory usage for attention but does not do much for speed. 

The way to get things to run faster would be to write CUDA/Triton to speed things up.

**Comment by yichen-xie-ai** [comment]
> [@kongwanbianjinyu](https://github.com/kongwanbianjinyu) just a quick update. we provide Triton Kernels here https://github.com/NVlabs/HMAR to speedup training.

@Kumbong Thank you for sharing this triton kernel. However, I notice that it will cost additional GPU memory compared with the default PyTorch implementation. Do you think it is normal or I made something wrong?

**Comment by Kumbong** [comment]
@yichen-xie-ai what are you benchmarking GPU are you benchmarking on ? what numbers are you seeing ?

**Comment by yichen-xie-ai** [comment]
@Kumbong Thank you for reply. I am using A100 GPUs. I try to run the training code with 1024x1024 images and batch size 4 on a single GPU using FP16. The following is the Default PyTorch and your triton implementation results:

PyTorch: Memory 56GB, Batch time 13.7s: 
Triton: Memory 61GB, Batch time 9.1s

It is great that a speed-up is witnessed, but the GPU memory actually increases.
